### ISSUE TITLE
Feature request: validate token on passthrough routes

### ISSUE THREAD

**Issue by masadelante** [tuupola/slim-jwt-auth#47]
I recently ran into a scenario where it would be handy to have decode a JWT token for anonymous routes, if it's available. In the API I'm constructing, I'm using strict rate limiting for some anynomous calls, and rate limits are raised when a user's logged in.

However, the middleware doesn't allow this at the moment, since it's an all or nothing proposition: either the route has to be authenticated, and an 401 is returned when no token is present, or a route is passthrough, and no token decoding is done at all.

I solved the problem for the moment by decoding the token in middleware that follows, but in my opinion, it would be more elegant if this could be done in this middleware.

Would you consider adding either a parameter specifying that passthrough routes get to be decoded too - without returning a 401, or alternatively to specify a third option apart from "path" and "passthrough" which allows routes to be used either anonymously or as an authenticated user?

Another option, I presume, would be to define all routes as authenticated, and allow those routes that may be anonymous to proceed without a 401 status, using the error callback, but that seems a bit kludgy to me. It also would involve rewriting the error logic, since the error handler isn't able to override the 401 status at the moment.

**Comment by hannrais** [comment]
The feature request is pretty old already, but I would still love to see this happening. My use case is to support both REST and GraphQL endpoints. In REST side this MW does good job, but having only single GraphQL endpoint I have to validate the possibly included token separately.

**Comment by legoguy1000** [comment]
I too would love this.  For my project there are endpoints that do not require authentication but if you are logged in, i'll present more information.

**Comment by aldas** [comment]
Would like to see it also.  Maybe just refactor this block
https://github.com/tuupola/slim-jwt-auth/blob/3.x/src/JwtAuthentication.php#L139
```
            $token = $this->fetchToken($request);
            $decoded = $this->decodeToken($token);
```
into protected method that we could override and add custom logic (ie. returning generic decoded token for public route if it is missing)

**Comment by tuupola** [comment]
I do not think this middleware should have responsibility of to conditionally authenticate if token exists and allow unauthenticated access if token does not exist. I also think this feature would be a big foot gun. 

However, to conditionally execute this middleware only when request has a token you could use [northwoods/conditional-middleware](https://github.com/northwoods/conditional-middleware).
